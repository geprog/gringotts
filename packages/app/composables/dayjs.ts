import dayjs from 'dayjs';

export function formatTime(date: Date) {
  return dayjs(date).format('HH:mm');
}

export function formatDate(date: Date) {
  return dayjs(date).format('DD.MM.YYYY');
}

export function formatDateTime(date: Date) {
  return dayjs(date).format('DD.MM.YYYY HH:mm');
}
