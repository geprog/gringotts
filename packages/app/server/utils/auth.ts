import type { H3Event, SessionConfig } from 'h3';

type User = {
  token: string;
  backendUrl: string;
};

const sessionConfig: SessionConfig = useRuntimeConfig().auth || {};

export type AuthSession = {
  token: string;
};

export async function useAuthSession(event: H3Event) {
  const session = await useSession<AuthSession>(event, sessionConfig);
  return session;
}

export async function getUser(event: H3Event): Promise<User | undefined> {
  const session = await useAuthSession(event);
  if (!session.data?.token) {
    return undefined;
  }

  const { baseUrl } = useRuntimeConfig().api;
  return {
    token: session.data.token,
    backendUrl: baseUrl,
  };
}

export async function requireUser(event: H3Event): Promise<User> {
  const user = await getUser(event);
  if (!user) {
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
    });
  }

  return user;
}

export async function getSessionHeader(event: H3Event) {
  const config = useRuntimeConfig();

  const sessionName = config.auth.name || 'h3';

  let sealedSession: string | undefined;

  // Try header first
  if (config.sessionHeader !== false) {
    const headerName =
      typeof config.sessionHeader === 'string'
        ? config.sessionHeader.toLowerCase()
        : `x-${sessionName.toLowerCase()}-session`;
    const headerValue = event.node.req.headers[headerName];
    if (typeof headerValue === 'string') {
      sealedSession = headerValue;
    }
  }

  // Fallback to cookies
  if (!sealedSession) {
    sealedSession = getCookie(event, sessionName);
  }

  if (!sealedSession) {
    throw createError({
      statusCode: 401,
      statusMessage: 'Unauthorized',
    });
  }

  return { [`x-${sessionName.toLowerCase()}-session`]: sealedSession };
}
