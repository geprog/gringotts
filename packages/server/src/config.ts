import dotenv from 'dotenv';
import path from 'path';

dotenv.config({
  path: path.join(__dirname, '..', '..', '..', '.env'),
});

const defaultPort = 3000;

export const config = {
  port: process.env.PORT ? parseInt(process.env.PORT) : defaultPort,
  publicUrl: process.env.PUBLIC_URL || `http://localhost:${defaultPort}`,
  postgresUrl: process.env.POSTGRES_URL as string,
  adminToken: process.env.ADMIN_TOKEN as string,
  jwtSecret: process.env.JWT_SECRET as string,
  gotenbergUrl: process.env.GOTENBERG_URL || 'http://localhost:3000',
  dataPath: process.env.STORAGE_PATH || path.join(__dirname, 'data'),
};

export function checkConfig(): void {
  const _config = config as Partial<typeof config>;

  if (!_config.publicUrl) {
    throw new Error('Please configure PUBLIC_URL');
  }

  if (!_config.postgresUrl) {
    throw new Error('Please configure POSTGRES_URL');
  }

  if (!_config.adminToken) {
    throw new Error('Please configure ADMIN_TOKEN');
  }
}
