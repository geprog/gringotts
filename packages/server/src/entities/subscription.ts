import { Collection, EntitySchema, ReferenceType } from '@mikro-orm/core';
import { v4 } from 'uuid';

import { Customer } from '~/entities/customer';
import { Invoice } from '~/entities/invoice';
import { Project } from '~/entities/project';
import { SubscriptionChange } from '~/entities/subscription_change';
import { SubscriptionPeriod } from '~/entities/subscription_period';

export class Subscription {
  _id: string = v4();
  anchorDate!: Date; // first date a user ever started a subscription for the object
  status: 'processing' | 'active' | 'error' = 'active';
  error?: string;
  lastPayment?: Date;
  currentPeriodStart!: Date;
  currentPeriodEnd!: Date;
  customer!: Customer;
  changes = new Collection<SubscriptionChange>(this);
  createdAt: Date = new Date();
  updatedAt: Date = new Date();
  invoices = new Collection<Invoice>(this);
  project!: Project;

  constructor(data?: Partial<Subscription>) {
    Object.assign(this, data);
  }

  /**
   * End the current plan and start with a new one
   * @param data.pricePerUnit Price per unit for the new plan
   * @param data.units Units for the new plan
   * @param data.changeDate Date when to end the current plan and start with a new one
   */
  changePlan(data: { pricePerUnit: number; units: number; changeDate?: Date }): void {
    // set end date of last change if we have one
    if (this.changes.count() > 0) {
      if (data.changeDate === undefined) {
        throw new Error('changeDate is required if you already have a change');
      }
      this.changes[this.changes.count() - 1].end = data.changeDate;
    }

    if (this.changes.getItems().filter((c) => c.end === undefined).length > 1) {
      throw new Error('Only the last item is allowed to have no end date');
    }

    this.changes.add(
      new SubscriptionChange({
        start: this.changes.count() === 0 ? this.anchorDate : (data.changeDate as Date),
        pricePerUnit: data.pricePerUnit,
        units: data.units,
        subscription: this,
      }),
    );
  }

  getPeriod(start: Date, end: Date): SubscriptionPeriod {
    return new SubscriptionPeriod(this, start, end);
  }

  toJSON(): Subscription {
    return {
      ...this,
      changes: this.changes?.isInitialized()
        ? this.changes.getItems().map((change) => ({ ...change, subscription: undefined }))
        : [],
    };
  }
}

export const subscriptionSchema = new EntitySchema<Subscription>({
  class: Subscription,
  properties: {
    _id: { type: 'uuid', onCreate: () => v4(), primary: true },
    anchorDate: { type: Date },
    status: { type: 'string', default: 'active' },
    error: { type: 'string', nullable: true },
    lastPayment: { type: Date, nullable: true },
    currentPeriodStart: { type: Date },
    currentPeriodEnd: { type: Date },
    customer: {
      reference: ReferenceType.MANY_TO_ONE,
      entity: () => Customer,
    },
    changes: {
      reference: ReferenceType.ONE_TO_MANY,
      entity: () => SubscriptionChange,
      mappedBy: (change: SubscriptionChange) => change.subscription,
    },
    createdAt: { type: Date, onCreate: () => new Date() },
    updatedAt: { type: Date, onUpdate: () => new Date() },
    invoices: {
      reference: ReferenceType.ONE_TO_MANY,
      entity: () => Invoice,
      mappedBy: (invoice: Invoice) => invoice.subscription,
    },
    project: {
      reference: ReferenceType.MANY_TO_ONE,
      entity: () => Project,
    },
  },
});
