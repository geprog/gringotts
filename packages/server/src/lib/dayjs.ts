import dayjs from 'dayjs';
import dayjsCustomParseFormat from 'dayjs/plugin/customParseFormat';
import dayjsIsBetween from 'dayjs/plugin/isBetween';
import dayjsMinMax from 'dayjs/plugin/minMax';

dayjs.extend(dayjsCustomParseFormat);
dayjs.extend(dayjsMinMax);
dayjs.extend(dayjsIsBetween);

export default dayjs;

export function formatDate(date: Date, format = 'dd.mm.YYYY'): string {
  return dayjs(date).format(format);
}
