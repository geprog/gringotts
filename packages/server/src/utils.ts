import dayjs from '~/lib/dayjs';

export function getPeriodFromAnchorDate(someDateInPeriod: Date, anchorDate: Date): { start: Date; end: Date } {
  const startDate = Math.min(dayjs(someDateInPeriod).endOf('month').date(), dayjs(anchorDate).date());
  let start = dayjs(someDateInPeriod).set('date', startDate).startOf('day');

  if (start.isAfter(someDateInPeriod)) {
    // select previous period
    const newStart = start.subtract(1, 'month');
    let date = Math.max(newStart.date(), dayjs(anchorDate).date());
    date = Math.min(date, newStart.endOf('month').date());
    start = newStart.set('date', date);
  }

  const endDate = Math.min(start.add(1, 'month').endOf('month').date(), dayjs(anchorDate).date());
  const end = start.add(1, 'month').endOf('month').set('date', endDate).subtract(1, 'day').endOf('day');

  return { start: start.toDate(), end: end.toDate() };
}

export function getNextPeriod(someDateInPeriod: Date, anchorDate: Date): { start: Date; end: Date } {
  const { end } = getPeriodFromAnchorDate(someDateInPeriod, anchorDate);
  return getPeriodFromAnchorDate(dayjs(end).add(1, 'day').toDate(), anchorDate);
}

export function getPreviousPeriod(nextPayment: Date, anchorDate: Date): { start: Date; end: Date } {
  const { start } = getPeriodFromAnchorDate(nextPayment, anchorDate);
  return getPeriodFromAnchorDate(dayjs(start).subtract(1, 'day').toDate(), anchorDate);
}

export function getActiveUntilDate(oldActiveUntil: Date, anchorDate: Date): Date {
  const { end } = getPeriodFromAnchorDate(oldActiveUntil, anchorDate);
  return end;
}

export function getNextPaymentDate(currentNextPayment: Date, anchorDate: Date): Date {
  const { start } = getNextPeriod(currentNextPayment, anchorDate);
  return dayjs(start).add(1, 'hour').toDate(); // add short buffer
}
