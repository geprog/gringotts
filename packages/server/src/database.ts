import { MikroORM } from '@mikro-orm/core';
import { EntityManager, EntityRepository, PostgreSqlDriver } from '@mikro-orm/postgresql';

import { config } from '~/config';
import {
  Customer,
  customerSchema,
  Invoice,
  InvoiceItem,
  invoiceItemSchema,
  invoiceSchema,
  Payment,
  paymentSchema,
  Subscription,
  SubscriptionChange,
  subscriptionChangeSchema,
  subscriptionSchema,
} from '~/entities';
import { addExitHook } from '~/lib/exit_hooks';

export class Database {
  orm!: MikroORM;

  async init(): Promise<void> {
    if (!config.postgresUrl) {
      throw new Error('POSTGRES_URL is not set');
    }

    this.orm = await MikroORM.init<PostgreSqlDriver>({
      type: 'postgresql',
      clientUrl: config.postgresUrl,
      entities: [
        customerSchema,
        subscriptionSchema,
        subscriptionChangeSchema,
        paymentSchema,
        invoiceSchema,
        invoiceItemSchema,
      ],
      discovery: { disableDynamicFileAccess: true },
    });
  }

  async connect(): Promise<void> {
    await this.orm.connect();

    const generator = this.orm.getSchemaGenerator();
    await generator.updateSchema();

    addExitHook(() => this.orm.close());
  }

  get em(): EntityManager {
    return this.orm.em.fork() as EntityManager;
  }

  get subscriptions(): EntityRepository<Subscription> {
    return this.em.getRepository(Subscription);
  }

  get customers(): EntityRepository<Customer> {
    return this.em.getRepository(Customer);
  }

  get subscriptionChanges(): EntityRepository<SubscriptionChange> {
    return this.em.getRepository(SubscriptionChange);
  }

  get payments(): EntityRepository<Payment> {
    return this.em.getRepository(Payment);
  }

  get invoices(): EntityRepository<Invoice> {
    return this.em.getRepository(Invoice);
  }

  get invoiceItems(): EntityRepository<InvoiceItem> {
    return this.em.getRepository(InvoiceItem);
  }
}

export const database = new Database();
