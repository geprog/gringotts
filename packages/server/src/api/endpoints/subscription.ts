import { FastifyInstance } from 'fastify';

import { getProjectFromRequest } from '~/api/helpers';
import { database } from '~/database';
import { Subscription } from '~/entities';
import { getActiveUntilDate, getNextPaymentDate } from '~/utils';

export function subscriptionEndpoints(server: FastifyInstance): void {
  server.post('/subscription', {
    schema: {
      summary: 'Create a subscription',
      tags: ['subscription'],
      body: {
        type: 'object',
        required: ['pricePerUnit', 'units', 'customerId'],
        additionalProperties: false,
        properties: {
          pricePerUnit: { type: 'number' },
          units: { type: 'number' },
          customerId: { type: 'string' },
        },
      },
      response: {
        200: {
          $ref: 'Subscription',
        },
        400: {
          $ref: 'ErrorResponse',
        },
        404: {
          $ref: 'ErrorResponse',
        },
        500: {
          $ref: 'ErrorResponse',
        },
      },
    },
    handler: async (request, reply) => {
      const project = await getProjectFromRequest(request);

      const body = request.body as {
        pricePerUnit: number;
        units: number;
        redirectUrl: string;
        customerId: string;
      };

      if (body.units < 1) {
        return reply.code(400).send({
          error: 'Units must be greater than 0',
        });
      }

      if (body.pricePerUnit < 0) {
        return reply.code(400).send({
          error: 'Price per unit must be greater than 0',
        });
      }

      const customer = await database.customers.findOne(
        { _id: body.customerId, project },
        { populate: ['subscriptions', 'activePaymentMethod'] },
      );
      if (!customer) {
        return reply.code(404).send({
          error: 'Customer not found',
        });
      }

      if (!customer.activePaymentMethod) {
        return reply.code(400).send({
          error: 'Customer has no active payment method',
        });
      }

      const now = new Date();

      const subscription = new Subscription({
        anchorDate: now,
        status: 'active',
        nextPayment: getNextPaymentDate(now, now),
        customer,
        project,
      });

      subscription.changePlan({ units: body.units, pricePerUnit: body.pricePerUnit });

      await database.em.persistAndFlush([customer, subscription]);

      await reply.send(subscription.toJSON());
    },
  });

  server.patch('/subscription/:subscriptionId', {
    schema: {
      summary: 'Patch a subscription',
      tags: ['subscription'],
      params: {
        type: 'object',
        required: ['subscriptionId'],
        additionalProperties: false,
        properties: {
          subscriptionId: { type: 'string' },
        },
      },
      body: {
        type: 'object',
        required: ['pricePerUnit', 'units'],
        additionalProperties: false,
        properties: {
          pricePerUnit: { type: 'number' },
          units: { type: 'number' },
        },
      },
      response: {
        200: {
          $ref: 'SuccessResponse',
        },
        400: {
          $ref: 'ErrorResponse',
        },
        404: {
          $ref: 'ErrorResponse',
        },
      },
    },
    handler: async (request, reply) => {
      const project = await getProjectFromRequest(request);

      const body = request.body as { pricePerUnit: number; units: number };

      if (body.units < 1) {
        return reply.code(400).send({
          error: 'Units must be greater than 0',
        });
      }

      if (body.pricePerUnit < 0) {
        return reply.code(400).send({
          error: 'Price per unit must be greater than 0',
        });
      }

      const { subscriptionId } = request.params as { subscriptionId: string };

      const subscription = await database.subscriptions.findOne(
        { _id: subscriptionId, project },
        { populate: ['customer', 'changes'] },
      );
      if (!subscription) {
        return reply.code(404).send({ error: 'Subscription not found' });
      }

      subscription.changePlan({ pricePerUnit: body.pricePerUnit, units: body.units, changeDate: new Date() });
      await database.em.persistAndFlush(subscription);

      await reply.send({ ok: true });
    },
  });

  server.get('/subscription/:subscriptionId', {
    schema: {
      summary: 'Get a subscription',
      tags: ['subscription'],
      params: {
        type: 'object',
        required: ['subscriptionId'],
        additionalProperties: false,
        properties: {
          subscriptionId: { type: 'string' },
        },
      },
      response: {
        200: {
          $ref: 'Subscription',
        },
        404: {
          $ref: 'ErrorResponse',
        },
      },
    },
    handler: async (request, reply) => {
      const project = await getProjectFromRequest(request);

      const { subscriptionId } = request.params as { subscriptionId: string };

      const subscription = await database.subscriptions.findOne(
        { _id: subscriptionId, project },
        { populate: ['customer', 'changes'] },
      );
      if (!subscription) {
        return reply.code(404).send({ error: 'Subscription not found' });
      }

      const activeUntil = subscription.lastPayment
        ? getActiveUntilDate(subscription.lastPayment, subscription.anchorDate)
        : undefined;

      const _subscription = {
        ...subscription.toJSON(),
        activeUntil,
      };

      await reply.send(_subscription);
    },
  });

  server.get('/subscription/:subscriptionId/invoice', {
    schema: {
      summary: 'List all invoices of a subscription',
      tags: ['subscription', 'invoice'],
      params: {
        type: 'object',
        required: ['subscriptionId'],
        additionalProperties: false,
        properties: {
          subscriptionId: { type: 'string' },
        },
      },
      response: {
        200: {
          type: 'array',
          items: {
            $ref: 'Invoice',
          },
        },
        404: {
          $ref: 'ErrorResponse',
        },
      },
    },
    handler: async (request, reply) => {
      const project = await getProjectFromRequest(request);

      const { subscriptionId } = request.params as { subscriptionId: string };

      const subscription = await database.subscriptions.findOne({ _id: subscriptionId, project });
      if (!subscription) {
        return reply.code(404).send({ error: 'Subscription not found' });
      }

      const invoices = await database.invoices.find({ subscription, project }, { populate: ['items'] });

      await reply.send(invoices.map((invoice) => invoice.toJSON()));
    },
  });
}
