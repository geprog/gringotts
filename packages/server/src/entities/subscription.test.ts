import { beforeAll, describe, expect, it } from 'vitest';

import { config } from '~/config';
import { database } from '~/database';
import dayjs from '~/lib/dayjs';
import { getPeriodFromAnchorDate } from '~/utils';

import { Subscription } from './subscription';

describe('Subscription', () => {
  beforeAll(async () => {
    config.postgresUrl = 'postgres://postgres:postgres@localhost:5432/postgres'; // set to dummy value so we can init database
    await database.init();
  });

  it('should handle simple month to month periods', () => {
    // given
    const subscription = new Subscription({
      anchorDate: dayjs('2020-01-01').toDate(),
    });
    subscription.changePlan({ pricePerUnit: 1, units: 50 });

    // when
    const price = subscription.getPeriod(dayjs('2020-01-31').toDate()).getInvoice().getPrice();

    // then
    expect(price).toStrictEqual(50);
  });

  it('should handle middle of month to middle of next month periods', () => {
    // given
    const subscription = new Subscription({
      anchorDate: dayjs('2020-01-15').toDate(),
    });
    subscription.changePlan({ pricePerUnit: 1, units: 50 });

    // when
    const price = subscription.getPeriod(dayjs('2020-02-15').toDate()).getInvoice().getPrice();

    // then
    expect(price).toStrictEqual(50);
  });

  it('should handle multiple changes of units and prices', () => {
    // given
    const subscription = new Subscription({
      anchorDate: dayjs('2020-01-01').toDate(),
    });
    subscription.changePlan({ pricePerUnit: 1, units: 50 });
    subscription.changePlan({ pricePerUnit: 1.5, units: 50, changeDate: dayjs('2020-01-16').toDate() });
    subscription.changePlan({ pricePerUnit: 2, units: 50, changeDate: dayjs('2020-01-19').toDate() });

    // when
    const price = subscription.getPeriod(dayjs('2020-02-01').toDate()).getInvoice().getPrice();

    // then
    expect(price).toStrictEqual(100);
  });

  it('should handle same day changes', () => {
    // given
    const anchorDate = dayjs('2020-01-01').toDate();
    const subscription = new Subscription({
      anchorDate,
    });
    subscription.changePlan({ pricePerUnit: 1, units: 50 });
    const { end } = getPeriodFromAnchorDate(anchorDate, anchorDate);
    subscription.changePlan({ pricePerUnit: 5, units: 50, changeDate: dayjs(end).subtract(5, 'hours').toDate() });

    // when
    const subscriptionPeriod = subscription.getPeriod(dayjs('2020-01-31').toDate());
    const price = subscriptionPeriod.getInvoice().getPrice();

    // then
    expect(price).toStrictEqual(51.34);
  });
});
