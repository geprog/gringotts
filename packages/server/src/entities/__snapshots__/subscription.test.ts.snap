// Vitest Snapshot v1

exports[`Subscription > should generate nice string invoices 1`] = `
"Invoice from 01.01.2020 00:00 to 01.02.2020 00:00
	1: 01.01.2020 00:00 - 16.01.2020 00:00:
		15 days of 31 = 0.48%
		1$ * 50units = 50$
		0.48% * 50$ = 24.19$
	2: 16.01.2020 00:00 - 19.01.2020 00:00:
		3 days of 31 = 0.1%
		1.5$ * 50units = 75$
		0.1% * 75$ = 7.26$
	3: 19.01.2020 00:00 - 01.02.2020 00:00:
		13 days of 31 = 0.42%
		2$ * 50units = 100$
		0.42% * 100$ = 41.94$
Total: 73.39$"
`;
